pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,483 @@ func (c *Controller) processNextWorkItem() bool {
-		if err != nil {
-			log.Printf("ERROR: Could not get config: %s", err.Error())
+	// do not update the status if there's no change
+	if !equality.Semantic.DeepEqual(origTenant.Status, tenant.Status) {
+		if _, err := c.updateStatus(tenant); err != nil {
+			if !c.isInitialized(origTenant) && c.isInitialized(tenant) {
+				c.deleteTenantNamespace(tenant.Status.TenantNamespaceName, tenant, config)
+			}
 			return err
 		}
-		tenantRoleName := config.GetTenantRoleName()
+	}
 
-		//TODO: handle updateProgress errors
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressCreateNamespace)
-		namespaceName, err = c.createNamespace(tenant)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorContent)
-		}
-		log.Printf("Create namespace successful for %s", namespaceName)
+	if reconcileErr != nil {
+		return reconcileErr
+	}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressGetServiceAccount)
-		account, err = c.getServiceAccount(tenant, defaultServiceAccountName)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorInfra)
-		}
+	c.updateMetrics()
+	c.syncCount++
+	return nil
+}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressAddRoleBinding)
-		var roleBinding *v1beta1.RoleBinding
-		roleBinding, err = c.addRoleBinding(account, tenant, tenantRoleName)
-		if err != nil {
-			return c.handleError(tenant, err, api.TenantResultErrorInfra)
-		}
-		log.Printf("Created Role Binding '%s' in namespace '%s'", roleBinding.GetName(), namespaceName)
+func (c *Controller) isInitialized(tenant *api.Tenant) bool {
+	return tenant.Status.TenantNamespaceName != ""
+}
 
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressFinalize)
-		tenant.Status.Result = api.TenantResultSuccess
-		tenant.Status.Message = "Tenant namespace successfully prepared"
-		if tenant, err = c.updateStatus(tenant); err != nil {
-			return err
-		}
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressFinished)
-		log.Printf("Tenant preparation successful for %s", tenant.GetName())
+func (c *Controller) reconcile(config clientConfig, tenant *api.Tenant) (err error) {
+	if c.isInitialized(tenant) {
+		err = c.reconcileInitialized(config, tenant)
+	} else {
+		err = c.reconcileUninitialized(config, tenant)
 	}
-	c.updateMetrics()
-	c.syncCount++
+	return
+}
+
+func (c *Controller) reconcileUninitialized(config clientConfig, tenant *api.Tenant) error {
+	c.logPrintln(tenant, "tenant not initialized yet")
+
+	nsName, err := c.createTenantNamespace(config, tenant)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
+		tenant.Status.SetCondition(&knativeapis.Condition{
+			Type:    knativeapis.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  api.StatusReasonFailed,
+			Message: condMsg,
+		})
+		return err
+	}
+
+	_, err = c.reconcileTenantRoleBinding(tenant, nsName, config)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
```

Comment:
```suggestion
		condMsg := fmt.Sprintf("Failed to create the tenant role binding.")
```

Reply:
The role binding is part of the tenant namespace setup. Therefore, if the role binding could not be created, the namespace creation fails (and we retry from scratch next time). So the message fits.
Note that the message of the condition is intentionally high-level. For details operators should check the controller logs. Later we might use events on the resource object to publish failure details.

Reply:
Fixed in 85b310470.